File: backend/contracts.py
from __future__ import annotations

from dataclasses import asdict, dataclass, field
from typing import Any, Dict, List


@dataclass
class MetricTile:
    label: str
    value: str
    detail: str
    tone: str = "normal"


@dataclass
class TelemetryPair:
    label: str
    value: str


@dataclass
class SubsystemCard:
    id: str
    name: str
    state: str
    summary: str
    detail: str
    telemetry: List[TelemetryPair] = field(default_factory=list)


@dataclass
class TimelineEvent:
    time: str
    title: str
    detail: str
    tone: str = "normal"


@dataclass
class ActionCommand:
    id: str
    label: str
    command: str
    variant: str
    detail: str


@dataclass
class PreviewPoint:
    x: float
    y: float
    magnitude: float
    tracked: bool


@dataclass
class DetectionSnapshot:
    pipeline: str
    mode: str
    fps: float
    frameMs: float
    dotCount: int
    meanDisp: float
    maxDisp: float
    missingRatio: float
    referenceLocked: bool
    previewPoints: List[PreviewPoint] = field(default_factory=list)
    notes: List[str] = field(default_factory=list)


@dataclass
class DashboardSnapshot:
    generatedAt: str
    missionName: str
    systemState: str
    mode: str
    platform: str
    uptime: str
    headline: str
    summary: str
    metrics: List[MetricTile] = field(default_factory=list)
    subsystems: List[SubsystemCard] = field(default_factory=list)
    detections: DetectionSnapshot | None = None
    actions: List[ActionCommand] = field(default_factory=list)
    timeline: List[TimelineEvent] = field(default_factory=list)
    operatorNotes: List[str] = field(default_factory=list)

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)

File: single_folder_runtime/nextjs-gui/lib/contracts.ts
export type SystemState = "ready" | "tracking" | "fault" | "maintenance";
export type Tone = "normal" | "positive" | "warning";
export type SubsystemState = "online" | "idle" | "warning" | "offline";
export type ActionVariant = "primary" | "secondary" | "danger";

export interface MetricTile {
  label: string;
  value: string;
  detail: string;
  tone: Tone;
}

export interface TelemetryPair {
  label: string;
  value: string;
}

export interface SubsystemCard {
  id: string;
  name: string;
  state: SubsystemState;
  summary: string;
  detail: string;
  telemetry: TelemetryPair[];
}

export interface TimelineEvent {
  time: string;
  title: string;
  detail: string;
  tone: Tone;
}

export interface ActionCommand {
  id: string;
  label: string;
  command: string;
  variant: ActionVariant;
  detail: string;
}

export interface PreviewPoint {
  x: number;
  y: number;
  magnitude: number;
  tracked: boolean;
}

export interface DetectionSnapshot {
  pipeline: string;
  mode: string;
  fps: number;
  frameMs: number;
  dotCount: number;
  meanDisp: number;
  maxDisp: number;
  missingRatio: number;
  referenceLocked: boolean;
  previewPoints: PreviewPoint[];
  notes: string[];
}

export interface DashboardSnapshot {
  generatedAt: string;
  missionName: string;
  systemState: SystemState;
  mode: string;
  platform: string;
  uptime: string;
  headline: string;
  summary: string;
  metrics: MetricTile[];
  subsystems: SubsystemCard[];
  detections: DetectionSnapshot;
  actions: ActionCommand[];
  timeline: TimelineEvent[];
  operatorNotes: string[];
}

export interface CommandRequest {
  type: string;
  payload?: Record<string, unknown>;
}

export interface CommandResult {
  accepted: boolean;
  message: string;
  echoedType: string;
  generatedAt: string;
}
